File: logbook-spring-webflux/src/main/java/org/zalando/logbook/spring/webflux/ServerResponse.java
package org.zalando.logbook.spring.webflux;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.util.MimeType;
import org.zalando.logbook.HttpHeaders;
import org.zalando.logbook.HttpResponse;
import org.zalando.logbook.Origin;

import javax.annotation.Nullable;
import java.nio.charset.Charset;
import java.util.Optional;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import static java.nio.charset.StandardCharsets.UTF_8;

@RequiredArgsConstructor
final class ServerResponse implements HttpResponse {

    private final ServerHttpResponse response;

    private final AtomicReference<State> state = new AtomicReference<>(new State.Unbuffered());

    @Override
    public Origin getOrigin() {
        return Origin.LOCAL;
    }

    @Override
    public HttpHeaders getHeaders() {
        Map<String, List<String>> map = new LinkedHashMap<>();
        response.getHeaders().forEach((name, values) ->
            map.put(name, new ArrayList<>(values))
        );
        return HttpHeaders.of(map);
    }

    @Override
    public int getStatus() {
        return Optional.ofNullable(response.getStatusCode())
                .orElse(HttpStatus.OK)
                .value();
    }

    @Nullable
    @Override
    public String getContentType() {
        return Optional.ofNullable(response.getHeaders().getContentType())
                .map(MimeType::toString)
                .orElse(null);
    }

    @Override
    public Charset getCharset() {
        return Optional.ofNullable(response.getHeaders().getContentType())
                .map(MimeType::getCharset)
                .orElse(UTF_8);
    }

    @Override
    public HttpResponse withBody() {
        state.updateAndGet(State::with);
        return this;
    }

    @Override
    public HttpResponse withoutBody() {
        state.updateAndGet(State::without);
        return this;
    }

    boolean shouldBuffer() {
        return state.get() instanceof State.Offering;
    }

    void buffer(byte[] message) {
        state.updateAndGet(s -> s.buffer(message));
    }

    @Override
    public byte[] getBody() {
        return state.get().getBody();
    }
}

File: logbook-ktor-server/src/main/kotlin/org/zalando/logbook/server/ServerResponse.kt
@file:Suppress(
    "SimpleRedundantLet", // jacoco workaround
    "RETURN_TYPE_MISMATCH_ON_OVERRIDE", // nullable content type in Kotlin
)

package org.zalando.logbook.server

import io.ktor.http.ContentType
import io.ktor.http.ContentType.Companion.parse
import io.ktor.http.HttpHeaders.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.http.charset
import io.ktor.server.request.httpVersion
import io.ktor.server.response.ApplicationResponse
import io.ktor.util.toMap
import org.zalando.logbook.HttpHeaders
import org.zalando.logbook.HttpResponse
import org.zalando.logbook.Origin
import org.zalando.logbook.common.State
import java.nio.charset.Charset
import java.util.concurrent.atomic.AtomicReference
import kotlin.text.Charsets.UTF_8

internal class ServerResponse(
    private val response: ApplicationResponse,
    private val body: Any,
) : HttpResponse {
    private val state: AtomicReference<State> = AtomicReference(State.Unbuffered())

    override fun getProtocolVersion(): String = response.call.request.httpVersion
    override fun getOrigin(): Origin = Origin.LOCAL
    override fun getHeaders(): HttpHeaders = HttpHeaders.of(response.headers.allValues().toMap())
    override fun getContentType(): String? = response.contentType?.let { it.toString().substringBefore(";") }
    override fun getCharset(): Charset = response.contentType?.charset() ?: UTF_8
    override fun getStatus(): Int = when {
        response.status() != null -> response.status()!!.value
        body is HttpStatusCode -> body.value
        else -> 200
    }
    override fun withBody(): HttpResponse = apply { state.updateAndGet { it.with() } }
    override fun withoutBody(): HttpResponse = apply { state.updateAndGet { it.without() } }
    override fun getBody(): ByteArray = state.get().body
    internal fun buffer(bytes: ByteArray) = state.updateAndGet { it.buffer(bytes) }
    internal fun shouldBuffer(): Boolean = state.get() is State.Offering

    private val ApplicationResponse.contentType: ContentType? get() = headers[ContentType]?.let { parse(it) }
}
